src/index/DeltaTriples.cpp

```diff
@@ -0,0 +1,189 @@
+      triplesDeleted_.erase(triple);
+    }
+  });
+
+  std::vector<LocatedTripleHandles> handles =
+      locateAndAddTriples(std::move(cancellationHandle), triples, true);
+
+  AD_CORRECTNESS_CHECK(triples.size() == handles.size());
+  // TODO<qup42>: replace with std::views::zip in C++23
+  for (size_t i = 0; i < triples.size(); i++) {
+    triplesInserted_.insert({triples[i], handles[i]});
+  }
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::deleteTriples(
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    std::vector<IdTriple<0>> triples) {
+  LOG(DEBUG) << "Deleting " << triples.size()
+             << " triples (including idempotent triples)." << std::endl;
+  std::ranges::sort(triples);
+  auto [first, last] = std::ranges::unique(triples);
+  triples.erase(first, last);
+  std::erase_if(triples, [this](const IdTriple<0>& triple) {
+    return triplesDeleted_.contains(triple);
+  });
+  std::ranges::for_each(triples, [this](const IdTriple<0>& triple) {
+    auto handle = triplesInserted_.find(triple);
+    if (handle != triplesInserted_.end()) {
+      eraseTripleInAllPermutations(handle->second);
+      triplesInserted_.erase(triple);
+    }
+  });
+
+  std::vector<LocatedTripleHandles> handles =
+      locateAndAddTriples(std::move(cancellationHandle), triples, false);
+
+  AD_CORRECTNESS_CHECK(triples.size() == handles.size());
+  // TODO<qup42>: replace with std::views::zip in C++23
+  for (size_t i = 0; i < triples.size(); i++) {
+    triplesDeleted_.insert({triples[i], handles[i]});
+  }
+}
+
+// ____________________________________________________________________________
+const LocatedTriplesPerBlock& DeltaTriples::getLocatedTriplesPerBlock(
+    Permutation::Enum permutation) const {
+  switch (permutation) {
+    case Permutation::PSO:
+      return locatedTriplesPerBlockInPSO_;
+    case Permutation::POS:
+      return locatedTriplesPerBlockInPOS_;
+    case Permutation::SPO:
+      return locatedTriplesPerBlockInSPO_;
+    case Permutation::SOP:
+      return locatedTriplesPerBlockInSOP_;
+    case Permutation::OSP:
+      return locatedTriplesPerBlockInOSP_;
+    case Permutation::OPS:
+      return locatedTriplesPerBlockInOPS_;
+    default:
+      AD_FAIL();
+  }
+}
+
+// ____________________________________________________________________________
+LocatedTriplesPerBlock& DeltaTriples::getLocatedTriplesPerBlock(
+    Permutation::Enum permutation) {
+  switch (permutation) {
+    case Permutation::PSO:
+      return locatedTriplesPerBlockInPSO_;
+    case Permutation::POS:
+      return locatedTriplesPerBlockInPOS_;
+    case Permutation::SPO:
+      return locatedTriplesPerBlockInSPO_;
+    case Permutation::SOP:
+      return locatedTriplesPerBlockInSOP_;
+    case Permutation::OSP:
+      return locatedTriplesPerBlockInOSP_;
+    case Permutation::OPS:
+      return locatedTriplesPerBlockInOPS_;
+    default:
+      AD_FAIL();
```

Comment:
If you really need the non-const getter, you can factor out into a common template function (if a sufficient amount of code remains).